### ISSUE TITLE
Foundry crashes with `arithmetic operation overflow`

### ISSUE THREAD

**Issue by benesjan** [foundry-rs/foundry#3596]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (70f4fb5 2022-11-02T00:05:06.592613Z)

### What command(s) is the bug in?

forge test

### Operating System

Linux

### Describe the bug

Hello, after updating `forge-std` dependency foundry started crashing with `arithmetic operation overflow` error. The issue is occuring locally on my M1 Macbook as well as in a linux docker container in CI. [Here](https://github.com/AztecProtocol/aztec-connect-bridges/pull/269) is the code and [here](https://app.circleci.com/pipelines/github/AztecProtocol/aztec-connect-bridges/1510/workflows/a62bbc0c-908b-42a6-a13e-8211bb4ecb96/jobs/1646) is the log of a failing CI pipeline.

### Update
The bug is not related to `forge-std` (the `forge-std` update simply made it occur more frequently). The bug is caused by overflow in smart contract execution.

**Comment by mattsse** [comment]
I think this is a U256 overflow, @mds1 were there any changes re balance or similar?

**Comment by mattsse** [comment]
@benesjan I believe it's the deal cheatcode that overflows here https://github.com/AztecProtocol/aztec-connect-bridges/blob/f8ef3f8dc5a3c21046d20fa235de6271586cb1a1/src/test/bridges/erc4626/ERC4626.t.sol#L108

**Comment by mds1** [comment]
There's no changes in forge-std that I can think of that would cause. Though it seems to me like this is a forge error, not a forge-std error. From lines 87-89 of the linked CI job:

```
The application panicked (crashed).
Message:  arithmetic operation overflow
Location: /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/primitive-types-0.11.1/src/lib.rs:38
```

If this was a solidity/EVM overflow it wouldn't panic

**Comment by mattsse** [comment]
okay it's the bound here 

https://github.com/AztecProtocol/aztec-connect-bridges/blob/f8ef3f8dc5a3c21046d20fa235de6271586cb1a1/src/test/bridges/erc4626/ERC4626.t.sol#L105

if you change the max to `type(uint96).max` it succeeds, unclear why there's a U256 overflow,
there's no cheatcode involved perhaps this is an issue inside the evm

**Comment by benesjan** [comment]
It seems that the `ERC4626Test` is failing with overflow error and weirdly it causes forge to crash for me most of the time.  Tried to run it a few times and only once from ~8 runs forge didn't crash.

<img width="1760" alt="image" src="https://user-images.githubusercontent.com/13470840/199741690-4498d228-0cec-46b3-a0a2-7ef7c6434b92.png">

**Comment by mattsse** [comment]
maybe related to the uint96 value fuzzing
@joshieDo any ideas?

**Comment by mds1** [comment]
@benesjan have you narrowed down the cause of this, i.e. are you sure it's the forge-std update as mentioned in the original description vs. a certain forge release that introduced this?

**Comment by mattsse** [comment]
I've narrowed this down to these lines

https://github.com/AztecProtocol/aztec-connect-bridges/blob/f8ef3f8dc5a3c21046d20fa235de6271586cb1a1/src/test/bridges/erc4626/ERC4626.t.sol#L56-L57

at least if I comment out the entire `testFullFlow` test and those two lines then the panic goes away.

```
 forge test --match-test "testFullFlow" --match-contract "ERC4626"
```

**Comment by benesjan** [comment]
@mds1 it's not caused by the update but it seems the update makes it more frequent. I think that my test are simply failing with the overflow error and for some reason it causes forge to crash. Sometimes forge doesn't crash and I see that the tests fail with the overflow error. I tried compiling an older forge version (from October 17) and the issue occurs there as well.

**Comment by mattsse** [comment]
the randomness of this is likely related to the fuzz test function's inputs.

Me guess at this point is, that the state changes inside the prank are picked up by the fuzzer (which looks at state) and perhaps the uint96 strategy has a problem with those values.

**Comment by benesjan** [comment]
Hello, this `forge` bug is still occurring. I have an unrelated example. I introduced a bug in test in [this commit](https://github.com/AztecProtocol/aztec-connect-bridges/pull/281/commits/608eb366563511d1b86b14339fc8740674b082b0) which caused overflow in the tested smart contract. This overflow caused forge to crash. Log can be seen in our CI pipeline [here](https://app.circleci.com/pipelines/github/AztecProtocol/aztec-connect-bridges/1590/workflows/e4b5f958-2305-4b64-969d-bb617f766c21/jobs/1731).

**Comment by mattsse** [comment]
is it still the same test?

**Comment by benesjan** [comment]
No, it's a different test. Previously it was an ERC4626 bridge test and now it is a TroveBridge test. They are unrelated and don't share basically any code.

**Comment by mattsse** [comment]
sorry, I meant test function

**Comment by benesjan** [comment]
It's a different test function. I run locally the following command and it fails:
`forge test --fork-block-number 14970000 -m 'testRedistributionSuccessfulSwap' --fork-url https://mainnet.infura.io/v3/9928b52099854248b3a096be07a6b23c`

It is on [this PR](https://github.com/AztecProtocol/aztec-connect-bridges/pull/281)'s branch.

Thanks

**Comment by mattsse** [comment]
thanks,

no fuzz test this time, makes it easier to debug, will have a closer look.

**Comment by mattsse** [comment]
@benesjan found it after I spent some time debugging the test.

there was an edge with deal+transfer that was triggered on revert. fixed in #3767

**Comment by benesjan** [comment]
@mattsse Thanks a ton! This was haunting us for a while 😌 Will close the issue once it gets to a release and I test it.

**Comment by benesjan** [comment]
@mattsse Unfortunately it seems that the issue was not eliminated. The bug [occured](https://app.circleci.com/pipelines/github/AztecProtocol/aztec-connect-bridges/1600/workflows/e64aaa7b-a051-4919-88ac-e54887d6e62f/jobs/1745) in our CI pipeline again and there we always install the newest release.

**Comment by benesjan** [comment]
Hello @mattsse, while trying to reproduce the bug I've stumbled upon another one:
<img width="899" alt="image" src="https://user-images.githubusercontent.com/13470840/202009104-01b79b82-c04d-471f-b197-1f819e6ee916.png">

After dealing 5 ETH to an address owner's balance suddenly drops to 1 wei.

This is how you should be able to reproduce it:

Checkout the `janb/foundry-bug` branch and run:
`forge test -m testLiquidationFlowWithCollateralToClaim --fork-url https://mainnet.infura.io/v3/9928b52099854248b3a096be07a6b23c --fork-block-number 15977385 -vv`

I just executed foundryup so I am using the newest version.

**Comment by sprw121** [comment]
@mattsse 

I believe I maybe suffering from the same issue. I have it narrowed down to a much smaller, very simple example. Would it be useful to post that information here? I don't want to clog up the thread unnecessarily.

**Comment by mattsse** [comment]
> @mattsse
> 
> I believe I maybe suffering from the same issue. I have it narrowed down to a much smaller, very simple example. Would it be useful to post that information here? I don't want to clog up the thread unnecessarily.

yes please, ty!

**Comment by sprw121** [comment]
Below are the test file and the library where I get the overflow. It's the UniswapV3 FullMath library (https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/FullMath.sol), but I copy paste the one I have locally.

I verified that when I simulate a transaction in tenderly, I DO NOT get the same overflow (it errors for a different but expected reason). There is an internal call that replicates the test exactly in the second call to `computeSwapStep` I believe. The details for this tenderly simulation are:

Contract: https://polygonscan.com/address/0x45dda9cb7c25131df268515131f647d726f50608
Function: swap()
Address: <pick your own, doesn't matter>
bool: true
int256: 97425036274
uint256: 2153965902990871670710978874800773
bytes: 0x

BlockNumber: 33948395
TxIndex: 33948395
From: <shouldn't matter, use 0x2791bca1f2de4661ed88a30c99a7a9449aa84174>
Gas: 8000000

Test File:
```
// SPDX-License-Identifier: GPL-3.0-or-later
pragma solidity ^0.8.6;

import "@forge-std/src/Test.sol";

import '../src/libraries/FullMath.sol';

contract TestOverflow is Test {
    function testOverflow() public {
        uint256 a = 558630655357838830346177341558886063651162161152;
        uint256 b = 1077198369626470956922431708142;
        uint256 denomiator = 2155043101360498141667901306508915;

        uint x = FullMath.mulDivRoundingUp(a,b,denomiator);
    }
}
```

Library that oveflows:
```
// SPDX-License-Identifier: MIT

pragma solidity >=0.4.0 <0.9.0;

/// @title Contains 512-bit math functions
/// @notice Facilitates multiplication and division that can have overflow of an intermediate value without any loss of precision
/// @dev Handles "phantom overflow" i.e., allows multiplication and division where an intermediate value overflows 256 bits
library FullMath {
    /// @notice Calculates floor(a×b÷denominator) with full precision. Throws if result overflows a uint256 or denominator == 0
    /// @param a The multiplicand
    /// @param b The multiplier
    /// @param denominator The divisor
    /// @return result The 256-bit result
    /// @dev Credit to Remco Bloemen under MIT license https://xn--2-umb.com/21/muldiv
    function mulDiv(
        uint256 a,
        uint256 b,
        uint256 denominator
    ) internal view returns (uint256 result) {
        // 512-bit multiply [prod1 prod0] = a * b
        // Compute the product mod 2**256 and mod 2**256 - 1
        // then use the Chinese Remainder Theorem to reconstruct
        // the 512 bit result. The result is stored in two 256
        // variables such that product = prod1 * 2**256 + prod0
        uint256 prod0; // Least significant 256 bits of the product
        uint256 prod1; // Most significant 256 bits of the product
        assembly {
            let mm := mulmod(a, b, not(0))
            prod0 := mul(a, b)
            prod1 := sub(sub(mm, prod0), lt(mm, prod0))
        }

        // Handle non-overflow cases, 256 by 256 division
        if (prod1 == 0) {
            require(denominator > 0);
            assembly {
                result := div(prod0, denominator)
            }
            return result;
        }

        // Make sure the result is less than 2**256.
        // Also prevents denominator == 0
        require(denominator > prod1);

        ///////////////////////////////////////////////
        // 512 by 256 division.
        ///////////////////////////////////////////////

        // Make division exact by subtracting the remainder from [prod1 prod0]
        // Compute remainder using mulmod
        uint256 remainder;
        assembly {
            remainder := mulmod(a, b, denominator)
        }
        // Subtract 256 bit number from 512 bit number
        assembly {
            prod1 := sub(prod1, gt(remainder, prod0))
            prod0 := sub(prod0, remainder)
        }

        // Factor powers of two out of denominator
        // Compute largest power of two divisor of denominator.
        // Always >= 1.
        uint256 twos = uint256(-1*int256(denominator) & int256(denominator));
        // Divide denominator by power of two
        assembly {
            denominator := div(denominator, twos)
        }

        // Divide [prod1 prod0] by the factors of two
        assembly {
            prod0 := div(prod0, twos)
        }
        // Shift in bits from prod1 into prod0. For this we need
        // to flip `twos` such that it is 2**256 / twos.
        // If twos is zero, then it becomes one
        assembly {
            twos := add(div(sub(0, twos), twos), 1)
        }
        prod0 |= prod1 * twos;

        // Invert denominator mod 2**256
        // Now that denominator is an odd number, it has an inverse
        // modulo 2**256 such that denominator * inv = 1 mod 2**256.
        // Compute the inverse by starting with a seed that is correct
        // correct for four bits. That is, denominator * inv = 1 mod 2**4
        uint256 inv = (3 * denominator) ^ 2;
        // Now use Newton-Raphson iteration to improve the precision.
        // Thanks to Hensel's lifting lemma, this also works in modular
        // arithmetic, doubling the correct bits in each step.
        inv *= 2 - denominator * inv; // inverse mod 2**8
        inv *= 2 - denominator * inv; // inverse mod 2**16
        inv *= 2 - denominator * inv; // inverse mod 2**32
        inv *= 2 - denominator * inv; // inverse mod 2**64
        inv *= 2 - denominator * inv; // inverse mod 2**128
        inv *= 2 - denominator * inv; // inverse mod 2**256

        // Because the division is now exact we can divide by multiplying
        // with the modular inverse of denominator. This will give us the
        // correct result modulo 2**256. Since the precoditions guarantee
        // that the outcome is less than 2**256, this is the final result.
        // We don't need to compute the high bits of the result and prod1
        // is no longer required.
        result = prod0 * inv;
        return result;
    }

    /// @notice Calculates ceil(a×b÷denominator) with full precision. Throws if result overflows a uint256 or denominator == 0
    /// @param a The multiplicand
    /// @param b The multiplier
    /// @param denominator The divisor
    /// @return result The 256-bit result
    function mulDivRoundingUp(
        uint256 a,
        uint256 b,
        uint256 denominator
    ) internal view returns (uint256 result) {
        result = mulDiv(a, b, denominator);
        if (mulmod(a, b, denominator) > 0) {
            require(result < type(uint256).max);
            result++;
        }
    }
}
```

**Comment by sprw121** [comment]
I suppose this could also be a compiler bug due to the fact the local contract is solidity 0.8.13 and the on-chain contract is 0.7.6, but this seems unlikely.

**Comment by sprw121** [comment]
> > @mattsse
> > I believe I maybe suffering from the same issue. I have it narrowed down to a much smaller, very simple example. Would it be useful to post that information here? I don't want to clog up the thread unnecessarily.
> 
> yes please, ty!

Pasted above

**Comment by mattsse** [comment]
@benesjan it appears either #3688 or 3687 has fixed that, couldn't reproduce on master, but present on nightly

**Comment by mattsse** [comment]
@sprw121 does your code make use of `deal` cheatcode? of not then this is unrelated, could you please open a separate issue for this, and ideally create a repro repo so I can reproduce this more easily?

**Comment by sprw121** [comment]
> @sprw121 does your code make use of `deal` cheatcode? of not then this is unrelated, could you please open a separate issue for this, and ideally create a repro repo so I can reproduce this more easily?

No cheat code. Yes I can try to put up a repro repo. Let me know if you want me to delete / clean up the comments in this thread.

**Comment by benesjan** [comment]
Hello @mattsse, our CI is now passing without any crashes so you can probably close this issue. Thank you for fixing this 👍

**Comment by chrismin** [comment]
Hi @mattsse — wanted to share that I'm unfortunately running into a similar issue. As far as I can tell, if a test combines the following, it leads to intermittent crashing of foundry:
- using a fuzzed uint256 value with the `deal` cheatcode
- `expectRevert`

When I get a chance, I will try to create a small repro example!

**Comment by chrismin** [comment]
@benesjan, I'm curious if you had to do anything specific to resolve the issue in your code?
